[RFC] vim-patch:7.4.1700,7.4.2219

jamessan opened the pull request:
#### vim-patch:7.4.1700

Problem:    Equivalence classes are not properly tested.
Solution:   Add tests for multi-byte and latin1. Fix an error. (Owen Leibman)

https://github.com/vim/vim/commit/22e421549d54147d003f6444de007cb1d73f1d27

src/regexp.c changes weren't applied because they're specific to EBCDIC
handling, which has been dropped from nvim.

The latin1-specific tests were also removed since neovim intends to
remove the ability to have 'encoding' set to anything other than utf8.
#### vim-patch:7.4.2219

Problem:    Recursive call to substitute gets stuck in sandbox. (Nikolai
            Pavlov)
Solution:   Handle the recursive call. (Christian Brabandt, closes vim/vim#950)
            Add a test.

https://github.com/vim/vim/commit/f5a39447a8ebe162ee62caa2ee502cd0e65eecaa

Closes #5118

jamessan committed:
vim-patch:7.4.2219

Problem:    Recursive call to substitute gets stuck in sandbox. (Nikolai
            Pavlov)
Solution:   Handle the recursive call. (Christian Brabandt, closes vim/vim#950)
            Add a test.

https://github.com/vim/vim/commit/f5a39447a8ebe162ee62caa2ee502cd0e65eecaa

Closes #5118

jamessan committed:
vim-patch:7.4.1700

Problem:    Equivalence classes are not properly tested.
Solution:   Add tests for multi-byte and latin1. Fix an error. (Owen Leibman)

https://github.com/vim/vim/commit/22e421549d54147d003f6444de007cb1d73f1d27

src/regexp.c changes weren't applied because they're specific to EBCDIC
handling, which has been dropped from nvim.

The latin1-specific tests were also removed since neovim intends to
remove the ability to have 'encoding' set to anything other than utf8.

jamessan committed:
lint

bfredl commented on src/nvim/testdir/test_regexp_latin.vim:
```diff
@@ -0,0 +1,39 @@
+" Tests for regexp in latin1 encoding
```
beyond possibly testing `:scriptencoding` command, this doesn't seem to test anything that the utf8 test already tests? If the purpose was to test `enc=latin1` I'll suggest just to remove this test, like we did for the latin1 spell test.

jamessan replied:
It was also to simplify merging of future patches, since Bram has been adding different tests to both files.  However, I can drop the latin1 file and just put the unique tests in the utf8 file.

jamessan committed:
ex_cmds: Factor pieces of functionality out of do_sub

* sub_joining_lines: Optimization for :%s/\n//
* sub_grow_buf: Allocation of buffer to contain replacement text
* sub_parse_flags: Parse {flags} from :s command into subflags_T

Although this doesn't reduce do_sub's size enough to satisfy lint, it
covers the more straightforward pieces.

jamessan commented:
I've removed the latin1 test and finished some refactoring of do_sub, which was intended to satisfy the lint check for function length.  However, in the end I silenced the lint check because I was having a hard time finding more pieces that could easily be extracted.  Also, that's the tack that was taken in #5226 so I didn't want to add extra noise when it comes to merging with that.

justinmk commented on src/nvim/ex_cmds.h:
```diff
@@ -26,6 +26,25 @@ typedef struct {
+  bool do_list;         ///< list last line with subs
+  bool do_number;       ///< list last line with line nr
+  SubIgnoreType do_ic;  ///< ignore case flag
+} subflags_T;
```
we generally don't expose types if they aren't needed outside of a module.

jamessan replied:
I did that because the auto-generated `.h` needed the definition, but I guess I should have just put it above the include for that file.

justinmk commented on src/nvim/ex_cmds.c:
```diff
@@ -3510,19 +3529,25 @@ void do_sub(exarg_T *eap)
           sandbox++;
         }
+        // Save flags for recursion.  They can change for e.g.
+        // :s/^/\=execute("s#^#gn")
```
in the original patch this is

    :s/^/\=execute("s#^##gn")

justinmk commented on src/nvim/ex_cmds.c:
```diff
@@ -3510,19 +3585,26 @@ void do_sub(exarg_T *eap)
+                                  sub_firstlnum - regmatch.startpos[0].lnum,
+                                  sub, sub_firstline, false, p_magic, true);
+        // Don't keep flags set by a recursive call
+        subflags = subflags_save;
```
If I remove this line, the test doesn't fail.

jamessan replied:
Hmm, yeah, I'm seeing some other odd behavior too.  I'll investigate.

jamessan replied:
The same happens with Vim, and that's actually expected (now that I think it through).  The subflags changes are only partially related to the sandbox problem.  They're just ensuring that flags in recursive calls to `:substitute` don't affect other calls.

In the case of the test, `:s/^/\=execute("s#^##gn")/` causes `do_count` to still be in effect after the recursive `do_sub` call, which is what leads to the `sandbox--;` statement, so to truly test the fix you also need to remove the `if (sandbox > 0)` check.

justinmk replied:
So the test fails correctly in that case ?

jamessan replied:
Yup.

```sh
$ git clean -fxd src/nvim
$ vim src/nvim/ex_cmds.c
$ git diff
diff --git i/src/nvim/ex_cmds.c w/src/nvim/ex_cmds.c
index b769199..696bebd 100644
--- i/src/nvim/ex_cmds.c
+++ w/src/nvim/ex_cmds.c
@@ -3599,12 +3599,10 @@ void do_sub(exarg_T *eap)
                                   sub_firstlnum - regmatch.startpos[0].lnum,
                                   sub, sub_firstline, false, p_magic, true);
         // Don't keep flags set by a recursive call
-        subflags = subflags_save;
+        //subflags = subflags_save;
         if (subflags.do_count) {
           curbuf->b_p_ma = save_ma;
-          if (sandbox > 0) {
             sandbox--;
-          }
           goto skip;
         }
 
$ make
ninja  -C .deps
ninja: Entering directory `.deps'
[2/2] cd /home/jamessan/src/github.com/neovim/.deps && /usr/bin/cmake -E touch .third-party
ninja  -C build
ninja: Entering directory `build'
[40/43] Building C object src/nvim/CMakeFiles/nvim.dir/ex_cmds.c.o
../src/nvim/ex_cmds.c: In function ‘do_sub’:
../src/nvim/ex_cmds.c:3596:20: warning: unused variable ‘subflags_save’ [-Wunused-variable]
         subflags_T subflags_save = subflags;
                    ^~~~~~~~~~~~~
[43/43] cd /home/jamessan/src/github.com/neovim/build/runtime && /usr/bin/cmake -E copy_directory ...n/src/github.com/neovim/build/bin/nvim -u NONE -i NONE -e --headless -c helptags\ ++t\ doc -c quit
$ make -C src/nvim/testdir test_regexp_utf8.res
make: Entering directory '/home/jamessan/src/github.com/neovim/src/nvim/testdir'
echo 'set $_exitcode = -1\nrun\nif $_exitcode != -1\n  quit\nend' > .gdbinit
VIMRUNTIME=../../../runtime; export VIMRUNTIME;  ../../../build/bin/nvim -u unix.vim -U NONE --noplugin -u NONE -S runtest.vim test_regexp_utf8.vim
Makefile:178: recipe for target 'test_regexp_utf8.res' failed
make: *** [test_regexp_utf8.res] Error 1
make: Leaving directory '/home/jamessan/src/github.com/neovim/src/nvim/testdir'
```

justinmk commented on src/nvim/ex_cmds.c:
```diff
@@ -3510,19 +3585,26 @@ void do_sub(exarg_T *eap)
         }
-        /* get length of substitution part */
+        // Save flags for recursion.  They can change for e.g.
+        // :s/^/\=execute("s#^#gn")
```
in the original patch this is

    :s/^/\=execute("s#^##gn")

justinmk commented:
LGTM
